### ISSUE TITLE
[azservicebus] Timeout in ReceiveMessages

### ISSUE THREAD

**Issue by praminda** [Azure/azure-sdk-for-go#17017]
### Bug Report

- pkg: `github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus`
- SDK version: `v0.3.2`
- go version: `go 1.16.2`

- What happened?
We recently moved from `azure-service-bus-go` and implemented an asb client using `azure-sdk-for-go`. This implementation was working fine for few days and then started to give below timeout error frequently. A manual service restart seem to fix the issue temporarily, but after some time error appears again in the logs. Our implementation is supposed to retry retrieving messages indefinitely but it doesn't seem to recover from the error and getting the messages from the topic.
Pls note we are subscribing to multiple topics using multiple clients here.
You can find our client implementation [here](https://github.com/wso2/product-microgateway/blob/348cad1cd2dbedf32672234c4cdb1990462dfd3e/adapter/pkg/messaging/azure_listener.go#L70).
**Error:**

    ```
    Failed to receive messages from ASB. read tcp 1*.*.*.113:xxxxxx->**.*.*.88:xxxx: i/o timeout
    ```

- What did you expect or want to happen?
If something unexpected causes this timeout error we expect it to recover when retrying.

- Anything we should know about your environment.
Our service is running on AKS

**Comment by praminda** [comment]
Thanks for the fix @richardpark-msft. 
When we will be able to get the fixed pkg released?

**Comment by pubudu538** [comment]
Hi team, Can we get a release with the latest fixes?

**Comment by paul2u** [comment]
Thank you @richardpark-msft for checking this issue. Would you update us the latest status? Thank you so much for you help.

**Comment by richardpark-msft** [comment]
(sorry for the late response, was on vacation)

For this fix you can do a `go get` on the commit hash.

```bash
go get github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus@ad8d573a3cea4dd0f094550a838d56fc3ecbcddf
```

**Comment by pubudu538** [comment]
Hi @richardpark-msft , 

We did the version upgrade based on your above comment. But still we see the connectivity issues. This time we don't see any issues in the application side. The app stopped receiving messages from ASB. When we check the subscription, we can see the queued messages. Could you please help with this?

Thank you!

**Comment by pubudu538** [comment]
---- Update ----

We thought of doing upgrading the client sdk to [1] as it has fixed some of the issues. 

[1] - https://github.com/Azure/azure-sdk-for-go/releases/tag/sdk%2Fmessaging%2Fazservicebus%2Fv0.3.6

**Comment by richardpark-msft** [comment]
@pubudu538 - v0.3.6 definitely has more reliability fixes so it's worth upgrading - we recategorized some more errors as "connection-reset required". However, the error you're describing was expected to be fixed with the commit you installed. 

Do you mind installing the latest package (was just released yesterday) and turning on debug logging as described here? It can be a bit verbose but it'll give me a better idea of what I might be missing.

https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azservicebus/README.md#logging

This will help me pinpoint what's happening internally.

**Comment by pubudu538** [comment]
@richardpark-msft We will install the latest package and update the status.

**Comment by AmaliMatharaarachchi** [comment]
We have upgraded to v0.3.6. We observed that sometimes the server hangs on this line forever `receiver.CompleteMessage(context, message)` [1]. 
[1] https://github.com/Azure/azure-sdk-for-go/blob/79f8e1f4bbe3045dfed5d481e880a259c78fcd44/sdk/messaging/azservicebus/receiver.go#L305

**Comment by richardpark-msft** [comment]
@AmaliMatharaarachchi, I'm looking at that. I think that might be related to a bug where 410 (lock lost) is treated as retryable. It shouldn't hang forever but it will retry, which can appear to be a hang. I have a fix incoming for that.

**Comment by richardpark-msft** [comment]
@AmaliMatharaarachchi
> We have upgraded to v0.3.6. We observed that sometimes the server hangs on this line forever `receiver.CompleteMessage(context, message)` [1]. [1]
> 
> https://github.com/Azure/azure-sdk-for-go/blob/79f8e1f4bbe3045dfed5d481e880a259c78fcd44/sdk/messaging/azservicebus/receiver.go#L305

Do you have logging enabled for these times? If not, can you try logging as well? It'll help confirm if you're seeing the bug I just mentioned or if you have found something new:

https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azservicebus/README.md#troubleshooting

**Comment by richardpark-msft** [comment]
> @AmaliMatharaarachchi, I'm looking at that. I think that might be related to a bug where 410 (lock lost) is treated as retryable. It shouldn't hang forever but it will retry, which can appear to be a hang. I have a fix incoming for that.

The fix for that particular bug in coming in https://github.com/Azure/azure-sdk-for-go/pull/17382

**Comment by VirajSalaka** [comment]
> @AmaliMatharaarachchi
> 
> > We have upgraded to v0.3.6. We observed that sometimes the server hangs on this line forever `receiver.CompleteMessage(context, message)` [1]. [1]
> > https://github.com/Azure/azure-sdk-for-go/blob/79f8e1f4bbe3045dfed5d481e880a259c78fcd44/sdk/messaging/azservicebus/receiver.go#L305
> 
> Do you have logging enabled for these times? If not, can you try logging as well? It'll help confirm if you're seeing the bug I just mentioned or if you have found something new:
> 
> https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azservicebus/README.md#troubleshooting

@richardpark-msft Sorry, we did not have azure-sdk-for-go debug logs enabled by the time this error was occurred.

**Comment by richardpark-msft** [comment]
> > @AmaliMatharaarachchi
> > > We have upgraded to v0.3.6. We observed that sometimes the server hangs on this line forever `receiver.CompleteMessage(context, message)` [1]. [1]
> > > https://github.com/Azure/azure-sdk-for-go/blob/79f8e1f4bbe3045dfed5d481e880a259c78fcd44/sdk/messaging/azservicebus/receiver.go#L305
> > 
> > 
> > Do you have logging enabled for these times? If not, can you try logging as well? It'll help confirm if you're seeing the bug I just mentioned or if you have found something new:
> > https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azservicebus/README.md#troubleshooting
> 
> @richardpark-msft Sorry, we did not have azure-sdk-for-go debug logs enabled by the time this error was occurred.

That's okay, it's hard to have the logs on at just the right moment. :)

I've made a recent fix in #17382 where we were retrying on expired tokens. Due to the retry logic (it's exponential) it could appear to be a hang. It wouldn't be forever though so it might not be what you were seeing. 

If this is your issue this also means that the lock for that particular message was possibly expired - so the call would have failed, even with my fix applied. it's possible you'll want to consider calling Receiver.RenewLock() (to extend the length of the token) or increase the lock time at the queue level (ie, through the Azure Portal).

**Comment by tkent** [comment]
Just to chime in here - I also bumped into the exact same symptoms as @praminda. After running a small golang app to pickup messages for somewhere between 10-20 days, it just... stops working with the error messages similar to those originally posted. We're also running in AKS and restarting the pod will fix it.

In our case, the errors looked a little bit different. They were connection reset errors that looks like this: 

```
read tcp XXX.XXX.XXX.67:45406-> XXX.XXX.XXX.204:5671: read: connection reset by peer
```

We were using `v0.3.4`. We're going to upgrade to the current `v0.3.6` and wait for a few weeks to see if it's still happening.

@richardpark-msft - We'd also like to deploy a version of our solution that can be robust without a wait-and-see approach. Do you have any recommendations?

Our current thought is to periodically recreate the receiver.

**Comment by richardpark-msft** [comment]
Hi @tkent :)

There are two bugs in this issue:

1. Timeout is not handled properly (ie, results in an endless set of connection reset errors)
    
    This was fixed in 0.3.6 - the basic cause is that our underlying protocol lib caches errors and this particular error requires an entire connection reset (we were only doing a link-level reset). So to fix this particular problem you can just upgrade. We'll be doing a 0.3.7 release next week as well which will have some more fixes. We try to keep the changelog reasonably up to date so hopefully it makes it simpler to track which release has fixed/added what.

    https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azservicebus/CHANGELOG.md

3. CompleteMessage() hangs

    There was a bug, fixed in #17382, where we were doing retries on 410 errors (basically, lock lost). For plain message locks these aren't retriable (ie, it does nothing and will never just "become" valid again). This was resulting in us doing retries and delays underneath that could make the CompleteMessage() call appear to be hung. This will be released in 0.3.7 (next week).

> @richardpark-msft - We'd also like to deploy a version of our solution that can be robust without a wait-and-see approach. Do you have any recommendations?

I don't want to leave you in that situation for sure. We release on a monthly cadence and are always addressing bugs and working on improving our reliability (through testing we do on the backend, etc..). Part of this beta process is really running things through the wringer, and that's definitely shaking out issues.

Your best strategy is to keep up to date with the releases, using the changelog as your guide as to what fixes you'll be picking up. 

With that said, periodically recreating the receiver will "work", but it's just more code you'll probably want to throw away later since the library does this internally. We don't store or cache messages in the Receiver so shutting down when you're not in an active call won't lose any data.

**Comment by tkent** [comment]
@richardpark-msft 

Thanks for the details about what's going on, it's very helpful to have that clear explanation. For now, we've moved to  `v0.3.6` and have had it running for about a week. The first error you pointed out (which is what we kept hitting), hasn't popped up again. We'll keep an eye out for `0.3.7` and move to that when it comes out. 

I follow that this public beta is still maturing, but I believe the entire golang community is a rough spot when it comes to this service. As I understand it, your options are:

- Use the [now-abandoned golang sdk](https://github.com/Azure/azure-service-bus-go), which never made it past preview
- Use this golang sdk, still in preview
- Write your own SDK from scratch, using the Azure APIs.

If there is a fourth option I've missed, I'd really like to know about it. Perhaps we've been using preview SDKs with an alternative exists?
